### PR TITLE
[11.0] - Foreign currency payments - wrong reconciliation test case

### PR DESCRIPTION
**Pull request by JordiBForgeFlow**
Description of the issue/feature this PR addresses:
If you issue and get paid for an invoice in a different currency than the company, the full reconciliation should only occur when the invoice has been paid in full from the invoice currency perspective. Now, if the amounts in invoice and payment match from a company currency perspective, the invoice is incorrectly flagged as paid.

Current behavior before PR:

Desired behavior after PR is merged:
Does not implement any fix. Just adds a successful and a failing test scenarios.

Steps to reproduce:
FAILING SCENARIO:
Day 1: Invoice Cust/001 to customer (expressed in CC)
Market value of CC (day 1): 1 CC = $0.5
* Dr. 100 CC / $50 - Accounts receivable
* Cr. 100 CC / $50 - Revenue

 Day 2: Receive payment for half invoice Cust/001 (in CC)
Market value of CC (day 2): 1 CC = $1
Payment transaction:
* Dr. 50 CC / $50 - CC Bank
* Cr. 50 CC / $50 - Accounts Receivable

At this point the invoice Invoice Cust/001  is flagged as 'paid'.

It is expected that the invoice should keep open, and after the second payment has been received, it should be flagged as paid, and an exchange rate journal entry should be created with the gain/loss.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by jbeficent** [comment]
@qdp-odoo what do you think about this?

**Comment by qdp-odoo** [comment]
I think that I need a rope, a chair and a tree

**Comment by jbeficent** [comment]
@qdp-odoo :) you made me laugh so much seeing your answer. Interestingly it works perfectly in other cases, as I describe in the successful test.

File: addons/account/tests/__init__.py [hunk 1]
```diff
@@ -18,3 +18,4 @@
 from . import test_setup_bar
 from . import test_tax
 from . import test_templates_consistency
+from . import test_bad_reconciliation
```
Comment (qdp-odoo) [inline addons/account/tests/__init__.py]: test should be added directly in test_reconciliation.py

File: addons/account/tests/test_bad_reconciliation.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+        return res
+
+    def _check_account_balance(self, account):
+        balance = sum(self.env['account.move.line'].search(
+            [('account_id', '=', account.id)]).mapped('balance'))
+        return balance
+
+    def test_01(self):
+        """ Implements a successful scenario """
```
Comment (qdp-odoo) [inline addons/account/tests/test_bad_reconciliation.py]: what successful scenario? is it new compared to the tests in test_reconciliation.py? if yes, it should be
* moved there
* be more explicit on what it tests

File: addons/account/tests/test_bad_reconciliation.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+        # Payment transaction:
+        # * Dr. 50 CC / $25 - CC Bank (valued at market price
+        # at the time of receiving the coins)
+        # * Cr. 50 CC / $25 - Accounts Receivable
+        #####
+        self.env['res.currency.rate'].create({
+            'currency_id': self.currency_cc.id,
+            'name': time.strftime('%Y') + '-01-02',
+            'rate': 2,
+        })
```
Comment (qdp-odoo) [inline addons/account/tests/test_bad_reconciliation.py]: for the sake of Clarity: if the rate didn't change, don't re-create a res.currency.rate

File: addons/account/tests/test_bad_reconciliation.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        invoice_cust_001.register_payment(payment_move_line)
+        self.assertEqual(invoice_cust_001.state, 'open',
+                         'Invoice is in status %s' % invoice_cust_001.state)
+        #####
+        # Day 2: Receive payment for half invoice Cust/001 (in CC)
+        # -------------------------------------------------------
+        # Market value of CC (day 2): 1 CC = $1
+
+        # Payment transaction:
+        # * Dr. 50 CC / $25 - CC Bank (valued at market price
```
Comment (qdp-odoo) [inline addons/account/tests/test_bad_reconciliation.py]: wrong comment: 
* Dr. 50 CC / $**50** - CC Bank (valued at market price

File: addons/account/tests/test_bad_reconciliation.py (hunk 2 above)
Comment (qdp-odoo) [reply addons/account/tests/test_bad_reconciliation.py]: My opinion is: I think this use case is already tested, so it should be removed from the PR

**Comment by jbeficent** [comment]
@qdp-odoo I have fixed as per your comments. But cannot run the test locally. It skips this test in accounting. Perhaps I am doing something wrong here.

**Comment by qdp-odoo** [comment]
Alas, I tested a fix for that on v11 and it breaks another use case: when you register a payment in the company currency and reconcile it with an invoice in secondary currency, the system automatically record an amount in foreign currency in the payment (for easiness of processing... let's call that "technical reasons"). See the test test_statement_euro_invoice_usd_transaction_euro_full for full description.

So merging this PR (real-fixed commit is [here](https://github.com/odoo-dev/odoo/commit/2a7754425f23d1a13851dc2bff5e7bf1e4f4704b)  btw) and fixing that issue would be too much of a refactoring and it's just too risky for a stable version. 

We'll keep that in mind and try in master as we already have some ongoing refactoring on this area. I'll keep you posted

**Comment by qdp-odoo** [comment]
the test finally landed in saas-11.2, at revision bb861825cac4ce16bab42941cc0a4365ed86bcdf.

Its fix is done at the next commit, on e1306066026649acd388b49d7429337d87cf819a

thanks for the contribution!

**Pull request closed without merging**